docs/proposals/api-types-package-structure.md

```diff
@@ -0,0 +1,473 @@
+│   │   ├── defaulting
+│   │   │   ├── generated_defaulters.go
+│   │   │   └── generated_default_stubs.go
+│   │   ├── install
+│   │   │   └── install_all_resources.go
+│   │   ├── types.go
+│   │   └── validation
+│   │       ├── generated_validation.go
+│   │       └── manual_validation_overrides.go
+│   ├── v1beta1
+│   │   ├── conversion
+│   │   │   ├── generated_conversion.go
+│   │   │   └── manual_conversion_overrides.go
+│   │   ├── defaulting
+│   │   │   ├── generated_defaulters.go
+│   │   │   └── generated_default_stubs.go
+│   │   ├── install
+│   │   │   └── install_all_resources.go
+│   │   ├── types.go
+│   │   └── validation
+│   │       ├── generated_validation.go
+│   │       └── manual_validation_overrides.go
+│   └── v2alpha1
+│       ├── conversion
+│       │   ├── generated_conversion.go
+│       │   └── manual_conversion_overrides.go
+│       ├── defaulting
+│       │   ├── generated_defaulters.go
+│       │   └── generated_default_stubs.go
+│       ├── install
+│       │   └── install_all_resources.go
+│       ├── types.go
+│       └── validation
+│           ├── generated_validation.go
+│           └── manual_validation_overrides.go
+└── unversioned
+    ├── meta
+    │   └── meta.go
+    ├── resource
+    │   └── quantity.go
+    ├── time
+    │   └── time.go
+    └── types.go
+```
+
+### Package: `apis/`
+
+Multiple groups may live in the same tree. Each group gets its own directory
+under `apis/`. We'll assume an `apis/` directory of this structure in all tools.
+In the above example, we show one group, named 'group'.
+
+TBD: Either test code will live here, or no code.
+
+### Package: `apis/unversioned/`
+
+This package and sub-packages are allowed to be imported and used by any
+group/version/resource. There is no internal/versioned distinction. Types in
+this directory in the main kubernetes repository are reusable by anyone in any
+API; creators of 3rd party APIs should think carefully before creating their own
+unversioned package.
+
+### Package: `apis/group/`
+
+Multiple versions live in a group. Each version gets its own subdirectory.
+
+A `types.go` file containing an "internal" version of the group's objects also
+lives here.  These types are used as a destination/source for conversion
+functions. There may be types for several different resources in this `types.go`
+file.
+
+### Package: `apis/group/install`
+
+This package installs *all* of the versions of the group, meaning it imports
+each nested install/ package.
+
+### Package: `apis/group/v1`
+
+'v1' is taken to be the current stable version of this api.
+
+A `types.go` file containing the v1 version of the group's objects lives here.
+There may be types for several different resources in this `types.go` file.
+
+This package has tightly controlled imports. It's allowed to include the
+Kubernetes unversioned API types (to be moved in/under `pkg/apis/unversioned`)
+and little else. In particular, it *must not* import its parent `apis/group`.
```

Comment:
> @deads2k I agree that some external versions should be able to import some other external versions, but not arbitrarily so-- in particular, there must not be cycles. I would like to be even more restrictive, by e.g. making a "common" group and locating there types that other groups are allowed to reuse.

I think being restrictive of the groups that can be included ends up in a place where you're becoming a gatekeeper on "blessed" types.  I'd like to avoid doing that, but obviously "no cycles" is a rule that makes a lot of sense.